### ISSUE TITLE
Best practice for sharing blocks across several projects?

### ISSUE THREAD

**Issue by ToonTalk** [jmoenig/Snap--Build-Your-Own-Blocks#1702]
When I improve or fix a bug in a block used in several of my projects, I open each project, import the new version of the block(s), and save.  Is this the best one can do?

**Comment by brianharvey** [comment]
Yes.  One might imagine libraries that would be dynamically loaded when a project is loaded, instead of baked into the project, but that would have the disadvantage of making it impossible to change the API when improving a library.  As it is, old projects still have the old library, so there's no concern about backward compatibility.

**Comment by micahcowan** [comment]
It occurs to me that, as a compromise, you could have a feature where you select via checkbox which projects you would like to update block definitions for (and, perhaps, which block definitions to include for such update).

A nice, useful feature, but a decent amount of work without being a terribly high priority, probably. Don't think I'd personally find motivation to offer a pull request for something like this unless I ran into this issue myself, probably, and I imagine the core devs might feel the same way. Even if I did run into it, I'd probably hack a script together to alter the XML directly, outside of _Snap!_ itself (though re-saving to cloud and/or browser space would obviously remain a pain in that case).

**Comment by cycomachead** [comment]
> Yes. One might imagine libraries that would be dynamically loaded when a project is loaded, instead of baked into the project, but that would have the disadvantage of making it impossible to change the API when improving a library. As it is, old projects still have the old library, so there's no concern about backward compatibility.

At that point we might as well allow a project to declare dependencies and then libraries can have version numbers and then you could have an 'ecosystem' and then we're pretty much where every other "grown up" language is! Woohoo!
/s

Without proper APIs to access dependent projects we're stuck with only hacky/slow solutions. But really the notion of a "library" is the biggest problem with all of this -- Snap! can export (and of course, import) libraries, but without any management of them from within Snap<em>!</em>, we'd be solving the wrong problems first. :)

**Comment by brianharvey** [comment]
Yeah, I vote no on this, even though I can see how in principle it's a good idea.  I'm perhaps unduly influenced by all the time I've spent trying to run *nix binaries for which the needed shared library versions are no longer available anywhere.  (Or with /usr not mounted, which is an even less relevant reason for me to be afraid of this idea.)

**Comment by ToonTalk** [comment]
As micahcowan suggested you can have an optional _update_ libraries feature. Since it is _updating_ a library already incorporated in a project the project will run as before if the library can't be found.
